### PR TITLE
[dbus] add telemetry data definition and API

### PR DESCRIPTION
**Pull request by morningboata**
This change is to simplify adding new telemetry/data on the ot-br side by
- declaring the data definition so we can easily review, align and merge telemetry data in the e2e system.
- reusing the data API so feature owner doesn't need to re-implement and test the IPC API for each telemetry data.
- Add wpan_stats telemetry data definition and handler.

More telemetry data definition will be defined in the proto file in future changes.

### CODE REVIEW

**Review by xwang146 — COMMENTED** [review]
We don't have any test for this new API?

**Comment by morningboata** [comment]
The unit test is made in tests/dbus/test_dbus_client.cpp.

File: src/proto/thread_telemetry.proto [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+// | Field Range  | Logging From
+// | [1 - 500)    | Primary fields logged from OTBR-agent/OpenThread.
+// | [500-600)    | OTBR vendor fields logged from OTBR-agent/OpenThread.
+// | Other        | Reserved for now.
+// ---------------------------------------------------------------------------
+// Usage:
+// Delete metrics: do not directly delete metrics. Deprecate it instead.
+message TelemetryData {
+    // TODO: add telemetry for wpan, topo, border_router, rcp, and coex.
+    reserved 1, 2, 3, 4, 5, 6;
```
[the remark this answers is not in the document]
Comment (xwang146) [reply src/proto/thread_telemetry.proto:18]: I mean, why do we reserve 1-6 here?

[the remark this answers is not in the document]
Comment (morningboata) [reply src/proto/thread_telemetry.proto:18]: Reserve for wpan, topo, border_router, rcp, and coex metrics tags. Timeline wise the implementation is not used for now so it's not blocking.

File: tests/dbus/test_dbus_client.cpp [hunk 2]
```diff
@@ -239,6 +239,15 @@ void CheckNat64(ThreadApiDBus *aApi)
 #endif
 }
 
+void CheckTelemetryData(ThreadApiDBus *aApi)
+{
+    OTBR_UNUSED_VARIABLE(aApi);
+#if OTBR_ENABLE_FEATURE_FLAGS
+    std::vector<uint8_t> responseTelemetryDataBytes;
+    TEST_ASSERT(aApi->GetTelemetryData(responseTelemetryDataBytes) == OTBR_ERROR_NONE);
```
Comment (superwhd) [inline tests/dbus/test_dbus_client.cpp:247]:
```suggestion
    std::vector<uint8_t> responseTelemetryDataBytes;
    
    TEST_ASSERT(aApi->GetTelemetryData(responseTelemetryDataBytes) == OTBR_ERROR_NONE);
```

File: src/dbus/server/dbus_thread_object.cpp [hunk 3]
```diff
@@ -1410,6 +1413,25 @@ otError DBusThreadObject::GetDnssdCountersHandler(DBusMessageIter &aIter)
 #endif // OTBR_ENABLE_DNSSD_DISCOVERY_PROXY
 }
 
+otError DBusThreadObject::GetTelemetryDataHandler(DBusMessageIter &aIter)
+{
+#if OTBR_ENABLE_FEATURE_FLAGS
```
Comment (superwhd) [inline src/dbus/server/dbus_thread_object.cpp:1418]: I wonder how this is related to feature flags?

File: src/proto/thread_telemetry.proto [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+// Thread Telemetry data definition.
+// The field range for your data definition is determined as:
+// ---------------------------------------------------------------------------
+// | Field Range  | Logging From
+// | [1 - 500)    | Primary fields logged from OTBR-agent/OpenThread.
+// | [500-600)    | OTBR vendor fields logged from OTBR-agent/OpenThread.
+// | Other        | Reserved for now.
+// ---------------------------------------------------------------------------
+// Usage:
+// Delete metrics: do not directly delete metrics. Deprecate it instead.
```
Comment (superwhd) [inline src/proto/thread_telemetry.proto:15]: `metrics` seems not defined in the context. Maybe just refer to it by `field`?

File: src/dbus/server/dbus_thread_object.cpp (hunk 3 above)
Comment (morningboata) [reply src/dbus/server/dbus_thread_object.cpp:1418]: Done.

File: src/dbus/server/introspect.xml [hunk 5]
```diff
@@ -916,6 +916,11 @@
       <annotation name="org.freedesktop.DBus.Property.EmitsChangedSignal" value="false"/>
     </property>
 
+    <!-- TelemetryData: The Thread telemetry data (defined as proto/thread_telemetry.proto)
+      in binary form. -->
+    <property name="FeatureFlagListData" type="ay" access="read">
```
[the remark this answers is not in the document]
Comment (morningboata) [reply src/dbus/server/introspect.xml]: Done.

File: tests/dbus/test_dbus_client.cpp (hunk 2 above)
Comment (morningboata) [reply tests/dbus/test_dbus_client.cpp]: Done.

File: src/proto/thread_telemetry.proto (hunk 4 above)
Comment (morningboata) [reply src/proto/thread_telemetry.proto]: Done.

File: src/dbus/server/dbus_thread_object.cpp (hunk 3 above)
Comment (morningboata) [reply src/dbus/server/dbus_thread_object.cpp]: Added OTBR_ENABLE_TELEMETRY_DATA_API flag.

File: tests/dbus/test_dbus_client.cpp [hunk 6]
```diff
@@ -239,6 +239,16 @@ void CheckNat64(ThreadApiDBus *aApi)
 #endif
 }
 
+void CheckTelemetryData(ThreadApiDBus *aApi)
+{
+    OTBR_UNUSED_VARIABLE(aApi);
+#if OTBR_ENABLE_TELEMETRY_DATA_API
+    std::vector<uint8_t> responseTelemetryDataBytes;
+
```
Comment (superwhd) [inline tests/dbus/test_dbus_client.cpp:247]: I'm a bit concerned that the test case doesn't cover the situation that there's any payload in the message.

File: .github/workflows/border_router.yml [hunk 7]
```diff
@@ -99,7 +99,7 @@ jobs:
             cert_scripts: ./tests/scripts/thread-cert/backbone/*.py
             packet_verification: 1
           - name: "Border Router TREL with FEATURE_FLAG (avahi)"
-            otbr_options: "-DOT_DUA=ON -DOT_ECDSA=ON -DOT_MLR=ON -DOT_SERVICE=ON -DOT_SRP_SERVER=ON -DOTBR_COVERAGE=ON -DOTBR_DUA_ROUTING=ON -DOTBR_FEATURE_FLAGS=ON -DOTBR_TREL=ON -DOTBR_DNS_UPSTREAM_QUERY=ON"
+            otbr_options: "-DOT_DUA=ON -DOT_ECDSA=ON -DOT_MLR=ON -DOT_SERVICE=ON -DOT_SRP_SERVER=ON -DOTBR_COVERAGE=ON -DOTBR_DUA_ROUTING=ON -DOTBR_FEATURE_FLAGS=ON -DOTBR_TELEMETRY_DATA_API=ON -DOTBR_TREL=ON -DOTBR_DNS_UPSTREAM_QUERY=ON"
```
Comment (xuyirio) [inline .github/workflows/border_router.yml:102]: Why do we enable the flag in this TC but not others?

File: tests/dbus/test_dbus_client.cpp (hunk 6 above)
Comment (morningboata) [reply tests/dbus/test_dbus_client.cpp:250]: I added the wpan_stats proto payload and assert its content now in the test case.

File: .github/workflows/border_router.yml (hunk 7 above)
Comment (morningboata) [reply .github/workflows/border_router.yml:102]: I want to test the tests/dbus/test_dbus_client.cpp and this flag addition provides the coverage.

File: src/dbus/server/dbus_thread_object.cpp [hunk 8]
```diff
@@ -1410,6 +1456,92 @@ otError DBusThreadObject::GetDnssdCountersHandler(DBusMessageIter &aIter)
...
+        SuccessOrExit(error = otPlatRadioGetTransmitPower(threadHelper->GetInstance(), &radioTxPower));
+        wpanStats->set_radio_tx_power(radioTxPower);
+    }
+
+    {
+        const otMacCounters *linkCounters = otLinkGetCounters(threadHelper->GetInstance());
+        wpanStats->set_phy_rx(linkCounters->mRxTotal);
+        wpanStats->set_phy_tx(linkCounters->mTxTotal);
+        wpanStats->set_mac_unicast_rx(linkCounters->mRxUnicast);
+        wpanStats->set_mac_unicast_tx(linkCounters->mTxUnicast);
```
Comment (superwhd) [inline src/dbus/server/dbus_thread_object.cpp:1493]:
```suggestion
        const otMacCounters *linkCounters = otLinkGetCounters(threadHelper->GetInstance());
        
        wpanStats->set_phy_rx(linkCounters->mRxTotal);
        wpanStats->set_phy_tx(linkCounters->mTxTotal);
        wpanStats->set_mac_unicast_rx(linkCounters->mRxUnicast);
        wpanStats->set_mac_unicast_tx(linkCounters->mTxUnicast);
```

File: src/dbus/server/dbus_thread_object.cpp [hunk 9]
```diff
@@ -1410,6 +1456,92 @@ otError DBusThreadObject::GetDnssdCountersHandler(DBusMessageIter &aIter)
...
+        wpanStats->set_mac_rx_fail_unknown_neighbor(linkCounters->mRxErrUnknownNeighbor);
+        wpanStats->set_mac_rx_fail_invalid_src_addr(linkCounters->mRxErrInvalidSrcAddr);
+        wpanStats->set_mac_rx_fail_fcs(linkCounters->mRxErrFcs);
+        wpanStats->set_mac_rx_fail_other(linkCounters->mRxErrOther);
+    }
+
+    {
+        const otIpCounters *ipCounters = otThreadGetIp6Counters(threadHelper->GetInstance());
+        wpanStats->set_ip_tx_success(ipCounters->mTxSuccess);
+        wpanStats->set_ip_rx_success(ipCounters->mRxSuccess);
```
Comment (superwhd) [inline src/dbus/server/dbus_thread_object.cpp:1524]:
```suggestion
        const otIpCounters *ipCounters = otThreadGetIp6Counters(threadHelper->GetInstance());
        
        wpanStats->set_ip_tx_success(ipCounters->mTxSuccess);
        wpanStats->set_ip_rx_success(ipCounters->mRxSuccess);
```

File: src/dbus/server/dbus_thread_object.cpp [hunk 10]
```diff
@@ -1410,6 +1456,92 @@ otError DBusThreadObject::GetDnssdCountersHandler(DBusMessageIter &aIter)
...
+
+    auto threadHelper = mNcp->GetThreadHelper();
+
+    // Begin of WpanStats section.
+    auto wpanStats = telemetryData.mutable_wpan_stats();
+    {
+        otDeviceRole     role  = otThreadGetDeviceRole(threadHelper->GetInstance());
+        otLinkModeConfig otCfg = otThreadGetLinkMode(threadHelper->GetInstance());
+        wpanStats->set_node_type(TelemetryNodeTypeFromRoleAndLinkMode(role, otCfg));
+    }
```
Comment (superwhd) [inline src/dbus/server/dbus_thread_object.cpp:1473]:
```suggestion
    auto wpanStats = telemetryData.mutable_wpan_stats();
    
    {
        otDeviceRole     role  = otThreadGetDeviceRole(threadHelper->GetInstance());
        otLinkModeConfig otCfg = otThreadGetLinkMode(threadHelper->GetInstance());
        
        wpanStats->set_node_type(TelemetryNodeTypeFromRoleAndLinkMode(role, otCfg));
    }
```

File: src/dbus/server/dbus_thread_object.cpp [hunk 11]
```diff
@@ -1410,6 +1456,92 @@ otError DBusThreadObject::GetDnssdCountersHandler(DBusMessageIter &aIter)
...
+
+    wpanStats->set_channel(otLinkGetChannel(threadHelper->GetInstance()));
+
+    {
+        uint16_t ccaFailureRate = otLinkGetCcaFailureRate(threadHelper->GetInstance());
+        wpanStats->set_mac_cca_fail_rate(static_cast<float>(ccaFailureRate) / 0xffff);
+    }
```
Comment (superwhd) [inline src/dbus/server/dbus_thread_object.cpp:1480]:
```suggestion
    {
        uint16_t ccaFailureRate = otLinkGetCcaFailureRate(threadHelper->GetInstance());
        
        wpanStats->set_mac_cca_fail_rate(static_cast<float>(ccaFailureRate) / 0xffff);
    }
```

File: src/dbus/server/dbus_thread_object.cpp [hunk 12]
```diff
@@ -1410,6 +1456,92 @@ otError DBusThreadObject::GetDnssdCountersHandler(DBusMessageIter &aIter)
...
+
+    {
+        int8_t radioTxPower;
+        SuccessOrExit(error = otPlatRadioGetTransmitPower(threadHelper->GetInstance(), &radioTxPower));
+        wpanStats->set_radio_tx_power(radioTxPower);
```
Comment (superwhd) [inline src/dbus/server/dbus_thread_object.cpp:1485]:
```suggestion
        int8_t radioTxPower;
        
        SuccessOrExit(error = otPlatRadioGetTransmitPower(threadHelper->GetInstance(), &radioTxPower));
        wpanStats->set_radio_tx_power(radioTxPower);
```

File: tests/dbus/test_dbus_client.cpp [hunk 13]
```diff
@@ -239,6 +242,34 @@ void CheckNat64(ThreadApiDBus *aApi)
...
+    TEST_ASSERT(aApi->GetTelemetryData(responseTelemetryDataBytes) == OTBR_ERROR_NONE);
+    // Print TelemetryData proto in hex format.
+    printf("TelemetryData bytes in hex: ");
+    for (uint8_t byte : responseTelemetryDataBytes)
+    {
+        printf("%02x ", byte);
+    }
+    printf("\n");
+    threadnetwork::TelemetryData telemetryData;
```
Comment (superwhd) [inline tests/dbus/test_dbus_client.cpp:259]: Move up to get together with other variable definitions.

**Comment by superwhd** [comment]
@zhanglongxia @Irving-cl I wonder if you have any comments on the implementation of the `WpanStats`.

File: src/dbus/server/dbus_thread_object.cpp [hunk 14]
```diff
@@ -117,6 +120,47 @@ static uint64_t ConvertOpenThreadUint64(const uint8_t *aValue)
     return val;
 }
 
+#if OTBR_ENABLE_TELEMETRY_DATA_API
+static uint32_t TelemetryNodeTypeFromRoleAndLinkMode(const otDeviceRole &aRole, const otLinkModeConfig &aLinkModeCfg)
+{
+    if (aRole == OT_DEVICE_ROLE_DISABLED)
```
[the remark this answers is not in the document]
Comment (morningboata) [reply src/dbus/server/dbus_thread_object.cpp:126]: Done, changed to switch case.

File: src/dbus/server/dbus_thread_object.cpp [hunk 15]
```diff
@@ -117,6 +120,47 @@ static uint64_t ConvertOpenThreadUint64(const uint8_t *aValue)
...
+        return threadnetwork::TelemetryData::NODE_TYPE_LEADER;
+    }
+    else if (aRole == OT_DEVICE_ROLE_CHILD)
+    {
+        if (!aLinkModeCfg.mRxOnWhenIdle)
+        {
+            return threadnetwork::TelemetryData::NODE_TYPE_SLEEPY_END;
+        }
+        else if (!aLinkModeCfg.mDeviceType)
+        { // If it's not a FTD, return as minimal end.
```
[the remark this answers is not in the document]
Comment (morningboata) [reply src/dbus/server/dbus_thread_object.cpp:149]: Done.

File: src/dbus/server/dbus_thread_object.cpp (hunk 8 above)
Comment (morningboata) [reply src/dbus/server/dbus_thread_object.cpp:1493]: Done.

File: src/dbus/server/dbus_thread_object.cpp (hunk 9 above)
Comment (morningboata) [reply src/dbus/server/dbus_thread_object.cpp:1524]: Done.

File: src/dbus/server/dbus_thread_object.cpp (hunk 10 above)
Comment (morningboata) [reply src/dbus/server/dbus_thread_object.cpp:1473]: Done.

File: src/dbus/server/dbus_thread_object.cpp (hunk 11 above)
Comment (morningboata) [reply src/dbus/server/dbus_thread_object.cpp:1480]: Done.

File: src/dbus/server/dbus_thread_object.cpp (hunk 12 above)
Comment (morningboata) [reply src/dbus/server/dbus_thread_object.cpp:1485]: Done.

File: tests/dbus/test_dbus_client.cpp (hunk 13 above)
Comment (morningboata) [reply tests/dbus/test_dbus_client.cpp]: Done.

File: src/dbus/server/dbus_thread_object.cpp [hunk 16]
```diff
@@ -1410,6 +1447,97 @@ otError DBusThreadObject::GetDnssdCountersHandler(DBusMessageIter &aIter)
...
+#if OTBR_ENABLE_TELEMETRY_DATA_API
+    otError                      error = OT_ERROR_NONE;
+    threadnetwork::TelemetryData telemetryData;
+
+    auto threadHelper = mNcp->GetThreadHelper();
+
+    // Begin of WpanStats section.
+    auto wpanStats = telemetryData.mutable_wpan_stats();
+    {
+        otDeviceRole     role  = otThreadGetDeviceRole(threadHelper->GetInstance());
```
Comment (superwhd) [inline src/dbus/server/dbus_thread_object.cpp:1461]:
```suggestion
    auto wpanStats = telemetryData.mutable_wpan_stats();

    {
        otDeviceRole     role  = otThreadGetDeviceRole(threadHelper->GetInstance());
```

**Review by superwhd — APPROVED** [review]
LGTM :+1:

File: src/dbus/server/dbus_thread_object.cpp (hunk 16 above)
Comment (morningboata) [reply src/dbus/server/dbus_thread_object.cpp:1461]: Done.

**Review by xuyirio — APPROVED** [review]

**Review by xwang146 — APPROVED** [review]

File: tests/dbus/test_dbus_client.cpp [hunk 17]
```diff
@@ -350,6 +381,7 @@ int main()
                             CheckMdnsInfo(api.get());
                             CheckDnssdCounters(api.get());
                             CheckNat64(api.get());
+                            CheckTelemetryData(api.get());
```
Comment (bukepo) [inline tests/dbus/test_dbus_client.cpp:384]:
```suggestion
#if OTBR_ENABLE_TELEMETRY_DATA_API
                            CheckTelemetryData(api.get());
#endif
```

File: tests/dbus/test_dbus_client.cpp [hunk 18]
```diff
@@ -239,6 +242,34 @@ void CheckNat64(ThreadApiDBus *aApi)
 #endif
 }
 
+void CheckTelemetryData(ThreadApiDBus *aApi)
+{
+    OTBR_UNUSED_VARIABLE(aApi);
+#if OTBR_ENABLE_TELEMETRY_DATA_API
```
Comment (bukepo) [inline tests/dbus/test_dbus_client.cpp:248]:
```suggestion
#if OTBR_ENABLE_TELEMETRY_DATA_API
void CheckTelemetryData(ThreadApiDBus *aApi)
{
```

**Review by bukepo — APPROVED** [review]

File: tests/dbus/test_dbus_client.cpp (hunk 17 above)
Comment (morningboata) [reply tests/dbus/test_dbus_client.cpp:384]: Done.

File: tests/dbus/test_dbus_client.cpp (hunk 18 above)
Comment (morningboata) [reply tests/dbus/test_dbus_client.cpp]: Done.

**Review by abtink — APPROVED** [review]
Thanks. LGTM.

File: src/dbus/server/dbus_thread_object.cpp [hunk 19]
```diff
@@ -117,6 +120,38 @@ static uint64_t ConvertOpenThreadUint64(const uint8_t *aValue)
     return val;
 }
 
+#if OTBR_ENABLE_TELEMETRY_DATA_API
+static uint32_t TelemetryNodeTypeFromRoleAndLinkMode(const otDeviceRole &aRole, const otLinkModeConfig &aLinkModeCfg)
+{
+    switch (aRole)
+    {
+    case OT_DEVICE_ROLE_DISABLED:
+        return threadnetwork::TelemetryData::NODE_TYPE_DISABLED;
```
Comment (jwhui) [inline src/dbus/server/dbus_thread_object.cpp:129]: code style nit: we prefer a single return statement at the end.

Comment (morningboata) [reply src/dbus/server/dbus_thread_object.cpp]: Done.

**Review by jwhui — APPROVED** [review]

**Pull request merged**